apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/RetryHandler.cs

```diff
@@ -0,0 +1,63 @@
+﻿// Copyright 2022 Google Inc. All Rights Reserved.
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Apis.Http;
+using Google.Apis.Storage.v1;
+using System;
+using System.Linq;
+using System.Net.Http;
+using System.Threading.Tasks;
+
+namespace Google.Cloud.Storage.V1
+{
+    /// <summary>
+    /// Handler for exponential backoff for request messages which permit retries.
+    /// This is similar to BackoffHandler in Google.Apis.Http, but that doesn't give us enough flexibility.
+    /// This handler does not enforce a maximum number of retries; that's configured via <see cref="ConfigurableMessageHandler.NumTries"/>.
+    /// </summary>
+    internal sealed class RetryHandler : IHttpUnsuccessfulResponseHandler
+    {
+        private static readonly int[] s_retriableErrorCodes = { 408, 429, 500, 502, 503, 504 };
+        private static RetryHandler s_instance = new RetryHandler();
+
+        private RetryHandler() { }
+
+        internal static void MarkAsRetriable<TResponse>(StorageBaseServiceRequest<TResponse> request) =>
```

Comment:
We'd also want to mark connection reset and broken stream errors as retryable. Have you been able to see what these will look like in dotnet? If not the testbench is a good way to test this out. https://github.com/googleapis/storage-testbench

Reply:
Hi Chris, yes I had discussed this with Jon as well and we decided to wait till we can test it with testbench to see if its needed for dotnet and not handled by default. I have documented this for now in the documentation : https://docs.google.com/document/d/14EjYUGrEkize4rZQ9j4PuMIrrPOyJNwYqhhWkrP8Cbw/edit?usp=sharing&resourcekey=0-4ks2t_XktfR0vL8Ak39REQ